<!DOCTYPE html>
<html lang="en">
<head>
    
    <meta charset="utf-8">
    <title>Home - Documentation</title>
    
    
    <script src="scripts/prettify/prettify.js"></script>
    <script src="scripts/prettify/lang-css.js"></script>
    <!--[if lt IE 9]>
      <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
    <link type="text/css" rel="stylesheet" href="styles/prettify.css">
    <link type="text/css" rel="stylesheet" href="styles/jsdoc.css">
    <script src="scripts/nav.js" defer></script>
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
</head>
<body>

<input type="checkbox" id="nav-trigger" class="nav-trigger" />
<label for="nav-trigger" class="navicon-button x">
  <div class="navicon"></div>
</label>

<label for="nav-trigger" class="overlay"></label>

<nav >
    
    <input type="text" id="nav-search" placeholder="Search" />
    
    <h2><a href="index.html">Home</a></h2><h2><a href="https://github.com/LTLA/bioconductor.js" target="_blank" class="menu-item" id="repository" >Code Repo</a></h2><h3>Classes</h3><ul><li><a href="Annotated.html">Annotated</a><ul class='methods'><li data-type='method' style='display: none;'><a href="Annotated.html#metadata">metadata</a></li><li data-type='method' style='display: none;'><a href="Annotated.html#setMetadata">setMetadata</a></li></ul></li><li><a href="DataFrame.html">DataFrame</a><ul class='methods'><li data-type='method' style='display: none;'><a href="DataFrame.html#column">column</a></li><li data-type='method' style='display: none;'><a href="DataFrame.html#columnNames">columnNames</a></li><li data-type='method' style='display: none;'><a href="DataFrame.html#hasColumn">hasColumn</a></li><li data-type='method' style='display: none;'><a href="DataFrame.html#metadata">metadata</a></li><li data-type='method' style='display: none;'><a href="DataFrame.html#numberOfColumns">numberOfColumns</a></li><li data-type='method' style='display: none;'><a href="DataFrame.html#numberOfRows">numberOfRows</a></li><li data-type='method' style='display: none;'><a href="DataFrame.html#removeColumn">removeColumn</a></li><li data-type='method' style='display: none;'><a href="DataFrame.html#rowNames">rowNames</a></li><li data-type='method' style='display: none;'><a href="DataFrame.html#setColumn">setColumn</a></li><li data-type='method' style='display: none;'><a href="DataFrame.html#setColumnNames">setColumnNames</a></li><li data-type='method' style='display: none;'><a href="DataFrame.html#setMetadata">setMetadata</a></li><li data-type='method' style='display: none;'><a href="DataFrame.html#setRowNames">setRowNames</a></li><li data-type='method' style='display: none;'><a href="DataFrame.html#sliceColumns">sliceColumns</a></li></ul></li><li><a href="DenseMatrix.html">DenseMatrix</a><ul class='methods'><li data-type='method' style='display: none;'><a href="DenseMatrix.html#column">column</a></li><li data-type='method' style='display: none;'><a href="DenseMatrix.html#isColumnMajor">isColumnMajor</a></li><li data-type='method' style='display: none;'><a href="DenseMatrix.html#numberOfColumns">numberOfColumns</a></li><li data-type='method' style='display: none;'><a href="DenseMatrix.html#numberOfRows">numberOfRows</a></li><li data-type='method' style='display: none;'><a href="DenseMatrix.html#row">row</a></li><li data-type='method' style='display: none;'><a href="DenseMatrix.html#setColumn">setColumn</a></li><li data-type='method' style='display: none;'><a href="DenseMatrix.html#setRow">setRow</a></li><li data-type='method' style='display: none;'><a href="DenseMatrix.html#setValues">setValues</a></li><li data-type='method' style='display: none;'><a href="DenseMatrix.html#values">values</a></li></ul></li><li><a href="GRanges.html">GRanges</a><ul class='methods'><li data-type='method' style='display: none;'><a href="GRanges.html#.empty">empty</a></li><li data-type='method' style='display: none;'><a href="GRanges.html#buildOverlapIndex">buildOverlapIndex</a></li><li data-type='method' style='display: none;'><a href="GRanges.html#elementMetadata">elementMetadata</a></li><li data-type='method' style='display: none;'><a href="GRanges.html#end">end</a></li><li data-type='method' style='display: none;'><a href="GRanges.html#metadata">metadata</a></li><li data-type='method' style='display: none;'><a href="GRanges.html#names">names</a></li><li data-type='method' style='display: none;'><a href="GRanges.html#ranges">ranges</a></li><li data-type='method' style='display: none;'><a href="GRanges.html#seqnames">seqnames</a></li><li data-type='method' style='display: none;'><a href="GRanges.html#setElementMetadata">setElementMetadata</a></li><li data-type='method' style='display: none;'><a href="GRanges.html#setMetadata">setMetadata</a></li><li data-type='method' style='display: none;'><a href="GRanges.html#setNames">setNames</a></li><li data-type='method' style='display: none;'><a href="GRanges.html#setRanges">setRanges</a></li><li data-type='method' style='display: none;'><a href="GRanges.html#setSeqnames">setSeqnames</a></li><li data-type='method' style='display: none;'><a href="GRanges.html#setStrand">setStrand</a></li><li data-type='method' style='display: none;'><a href="GRanges.html#start">start</a></li><li data-type='method' style='display: none;'><a href="GRanges.html#strand">strand</a></li><li data-type='method' style='display: none;'><a href="GRanges.html#width">width</a></li></ul></li><li><a href="GRangesOverlapIndex.html">GRangesOverlapIndex</a><ul class='methods'><li data-type='method' style='display: none;'><a href="GRangesOverlapIndex.html#overlap">overlap</a></li></ul></li><li><a href="GroupedGRanges.html">GroupedGRanges</a><ul class='methods'><li data-type='method' style='display: none;'><a href="GroupedGRanges.html#.empty">empty</a></li><li data-type='method' style='display: none;'><a href="GroupedGRanges.html#buildOverlapIndex">buildOverlapIndex</a></li><li data-type='method' style='display: none;'><a href="GroupedGRanges.html#elementMetadata">elementMetadata</a></li><li data-type='method' style='display: none;'><a href="GroupedGRanges.html#group">group</a></li><li data-type='method' style='display: none;'><a href="GroupedGRanges.html#metadata">metadata</a></li><li data-type='method' style='display: none;'><a href="GroupedGRanges.html#names">names</a></li><li data-type='method' style='display: none;'><a href="GroupedGRanges.html#numberOfGroups">numberOfGroups</a></li><li data-type='method' style='display: none;'><a href="GroupedGRanges.html#rangeLengths">rangeLengths</a></li><li data-type='method' style='display: none;'><a href="GroupedGRanges.html#ranges">ranges</a></li><li data-type='method' style='display: none;'><a href="GroupedGRanges.html#rangeStarts">rangeStarts</a></li><li data-type='method' style='display: none;'><a href="GroupedGRanges.html#setElementMetadata">setElementMetadata</a></li><li data-type='method' style='display: none;'><a href="GroupedGRanges.html#setGroup">setGroup</a></li><li data-type='method' style='display: none;'><a href="GroupedGRanges.html#setMetadata">setMetadata</a></li><li data-type='method' style='display: none;'><a href="GroupedGRanges.html#setNames">setNames</a></li><li data-type='method' style='display: none;'><a href="GroupedGRanges.html#setRanges">setRanges</a></li></ul></li><li><a href="GroupedGRangesOverlapIndex.html">GroupedGRangesOverlapIndex</a><ul class='methods'><li data-type='method' style='display: none;'><a href="GroupedGRangesOverlapIndex.html#overlap">overlap</a></li></ul></li><li><a href="IRanges.html">IRanges</a><ul class='methods'><li data-type='method' style='display: none;'><a href="IRanges.html#.empty">empty</a></li><li data-type='method' style='display: none;'><a href="IRanges.html#buildOverlapIndex">buildOverlapIndex</a></li><li data-type='method' style='display: none;'><a href="IRanges.html#elementMetadata">elementMetadata</a></li><li data-type='method' style='display: none;'><a href="IRanges.html#end">end</a></li><li data-type='method' style='display: none;'><a href="IRanges.html#metadata">metadata</a></li><li data-type='method' style='display: none;'><a href="IRanges.html#names">names</a></li><li data-type='method' style='display: none;'><a href="IRanges.html#setElementMetadata">setElementMetadata</a></li><li data-type='method' style='display: none;'><a href="IRanges.html#setMetadata">setMetadata</a></li><li data-type='method' style='display: none;'><a href="IRanges.html#setNames">setNames</a></li><li data-type='method' style='display: none;'><a href="IRanges.html#setStart">setStart</a></li><li data-type='method' style='display: none;'><a href="IRanges.html#setWidth">setWidth</a></li><li data-type='method' style='display: none;'><a href="IRanges.html#start">start</a></li><li data-type='method' style='display: none;'><a href="IRanges.html#width">width</a></li></ul></li><li><a href="IRangesOverlapIndex.html">IRangesOverlapIndex</a><ul class='methods'><li data-type='method' style='display: none;'><a href="IRangesOverlapIndex.html#overlap">overlap</a></li></ul></li><li><a href="List.html">List</a><ul class='methods'><li data-type='method' style='display: none;'><a href="List.html#delete">delete</a></li><li data-type='method' style='display: none;'><a href="List.html#deleteByIndex">deleteByIndex</a></li><li data-type='method' style='display: none;'><a href="List.html#deleteByName">deleteByName</a></li><li data-type='method' style='display: none;'><a href="List.html#get">get</a></li><li data-type='method' style='display: none;'><a href="List.html#getByIndex">getByIndex</a></li><li data-type='method' style='display: none;'><a href="List.html#getByName">getByName</a></li><li data-type='method' style='display: none;'><a href="List.html#has">has</a></li><li data-type='method' style='display: none;'><a href="List.html#length">length</a></li><li data-type='method' style='display: none;'><a href="List.html#names">names</a></li><li data-type='method' style='display: none;'><a href="List.html#nameToIndex">nameToIndex</a></li><li data-type='method' style='display: none;'><a href="List.html#set">set</a></li><li data-type='method' style='display: none;'><a href="List.html#setByIndex">setByIndex</a></li><li data-type='method' style='display: none;'><a href="List.html#setByName">setByName</a></li><li data-type='method' style='display: none;'><a href="List.html#setNames">setNames</a></li><li data-type='method' style='display: none;'><a href="List.html#sliceIndices">sliceIndices</a></li><li data-type='method' style='display: none;'><a href="List.html#sliceRange">sliceRange</a></li><li data-type='method' style='display: none;'><a href="List.html#values">values</a></li></ul></li><li><a href="RangedSummarizedExperiment.html">RangedSummarizedExperiment</a><ul class='methods'><li data-type='method' style='display: none;'><a href="RangedSummarizedExperiment.html#assay">assay</a></li><li data-type='method' style='display: none;'><a href="RangedSummarizedExperiment.html#assayNames">assayNames</a></li><li data-type='method' style='display: none;'><a href="RangedSummarizedExperiment.html#assays">assays</a></li><li data-type='method' style='display: none;'><a href="RangedSummarizedExperiment.html#columnData">columnData</a></li><li data-type='method' style='display: none;'><a href="RangedSummarizedExperiment.html#columnNames">columnNames</a></li><li data-type='method' style='display: none;'><a href="RangedSummarizedExperiment.html#metadata">metadata</a></li><li data-type='method' style='display: none;'><a href="RangedSummarizedExperiment.html#numberOfAssays">numberOfAssays</a></li><li data-type='method' style='display: none;'><a href="RangedSummarizedExperiment.html#numberOfColumns">numberOfColumns</a></li><li data-type='method' style='display: none;'><a href="RangedSummarizedExperiment.html#numberOfRows">numberOfRows</a></li><li data-type='method' style='display: none;'><a href="RangedSummarizedExperiment.html#removeAssay">removeAssay</a></li><li data-type='method' style='display: none;'><a href="RangedSummarizedExperiment.html#rowData">rowData</a></li><li data-type='method' style='display: none;'><a href="RangedSummarizedExperiment.html#rowNames">rowNames</a></li><li data-type='method' style='display: none;'><a href="RangedSummarizedExperiment.html#rowRanges">rowRanges</a></li><li data-type='method' style='display: none;'><a href="RangedSummarizedExperiment.html#setAssay">setAssay</a></li><li data-type='method' style='display: none;'><a href="RangedSummarizedExperiment.html#setAssayNames">setAssayNames</a></li><li data-type='method' style='display: none;'><a href="RangedSummarizedExperiment.html#setAssays">setAssays</a></li><li data-type='method' style='display: none;'><a href="RangedSummarizedExperiment.html#setColumnData">setColumnData</a></li><li data-type='method' style='display: none;'><a href="RangedSummarizedExperiment.html#setColumnNames">setColumnNames</a></li><li data-type='method' style='display: none;'><a href="RangedSummarizedExperiment.html#setMetadata">setMetadata</a></li><li data-type='method' style='display: none;'><a href="RangedSummarizedExperiment.html#setRowData">setRowData</a></li><li data-type='method' style='display: none;'><a href="RangedSummarizedExperiment.html#setRowNames">setRowNames</a></li><li data-type='method' style='display: none;'><a href="RangedSummarizedExperiment.html#setRowRanges">setRowRanges</a></li><li data-type='method' style='display: none;'><a href="RangedSummarizedExperiment.html#sliceAssays">sliceAssays</a></li></ul></li><li><a href="SingleCellExperiment.html">SingleCellExperiment</a><ul class='methods'><li data-type='method' style='display: none;'><a href="SingleCellExperiment.html#alternativeExperiment">alternativeExperiment</a></li><li data-type='method' style='display: none;'><a href="SingleCellExperiment.html#alternativeExperimentNames">alternativeExperimentNames</a></li><li data-type='method' style='display: none;'><a href="SingleCellExperiment.html#alternativeExperiments">alternativeExperiments</a></li><li data-type='method' style='display: none;'><a href="SingleCellExperiment.html#assay">assay</a></li><li data-type='method' style='display: none;'><a href="SingleCellExperiment.html#assayNames">assayNames</a></li><li data-type='method' style='display: none;'><a href="SingleCellExperiment.html#assays">assays</a></li><li data-type='method' style='display: none;'><a href="SingleCellExperiment.html#columnData">columnData</a></li><li data-type='method' style='display: none;'><a href="SingleCellExperiment.html#columnNames">columnNames</a></li><li data-type='method' style='display: none;'><a href="SingleCellExperiment.html#mainExperimentName">mainExperimentName</a></li><li data-type='method' style='display: none;'><a href="SingleCellExperiment.html#metadata">metadata</a></li><li data-type='method' style='display: none;'><a href="SingleCellExperiment.html#numberOfAssays">numberOfAssays</a></li><li data-type='method' style='display: none;'><a href="SingleCellExperiment.html#numberOfColumns">numberOfColumns</a></li><li data-type='method' style='display: none;'><a href="SingleCellExperiment.html#numberOfRows">numberOfRows</a></li><li data-type='method' style='display: none;'><a href="SingleCellExperiment.html#reducedDimension">reducedDimension</a></li><li data-type='method' style='display: none;'><a href="SingleCellExperiment.html#reducedDimensionNames">reducedDimensionNames</a></li><li data-type='method' style='display: none;'><a href="SingleCellExperiment.html#reducedDimensions">reducedDimensions</a></li><li data-type='method' style='display: none;'><a href="SingleCellExperiment.html#removeAlternativeExperiment">removeAlternativeExperiment</a></li><li data-type='method' style='display: none;'><a href="SingleCellExperiment.html#removeAssay">removeAssay</a></li><li data-type='method' style='display: none;'><a href="SingleCellExperiment.html#removeReducedDimension">removeReducedDimension</a></li><li data-type='method' style='display: none;'><a href="SingleCellExperiment.html#rowData">rowData</a></li><li data-type='method' style='display: none;'><a href="SingleCellExperiment.html#rowNames">rowNames</a></li><li data-type='method' style='display: none;'><a href="SingleCellExperiment.html#rowRanges">rowRanges</a></li><li data-type='method' style='display: none;'><a href="SingleCellExperiment.html#setAlternativeExperiment">setAlternativeExperiment</a></li><li data-type='method' style='display: none;'><a href="SingleCellExperiment.html#setAlternativeExperimentNames">setAlternativeExperimentNames</a></li><li data-type='method' style='display: none;'><a href="SingleCellExperiment.html#setAlternativeExperiments">setAlternativeExperiments</a></li><li data-type='method' style='display: none;'><a href="SingleCellExperiment.html#setAssay">setAssay</a></li><li data-type='method' style='display: none;'><a href="SingleCellExperiment.html#setAssayNames">setAssayNames</a></li><li data-type='method' style='display: none;'><a href="SingleCellExperiment.html#setAssays">setAssays</a></li><li data-type='method' style='display: none;'><a href="SingleCellExperiment.html#setColumnData">setColumnData</a></li><li data-type='method' style='display: none;'><a href="SingleCellExperiment.html#setColumnNames">setColumnNames</a></li><li data-type='method' style='display: none;'><a href="SingleCellExperiment.html#setMainExperimentName">setMainExperimentName</a></li><li data-type='method' style='display: none;'><a href="SingleCellExperiment.html#setMetadata">setMetadata</a></li><li data-type='method' style='display: none;'><a href="SingleCellExperiment.html#setReducedDimension">setReducedDimension</a></li><li data-type='method' style='display: none;'><a href="SingleCellExperiment.html#setReducedDimensionNames">setReducedDimensionNames</a></li><li data-type='method' style='display: none;'><a href="SingleCellExperiment.html#setReducedDimensions">setReducedDimensions</a></li><li data-type='method' style='display: none;'><a href="SingleCellExperiment.html#setRowData">setRowData</a></li><li data-type='method' style='display: none;'><a href="SingleCellExperiment.html#setRowNames">setRowNames</a></li><li data-type='method' style='display: none;'><a href="SingleCellExperiment.html#setRowRanges">setRowRanges</a></li><li data-type='method' style='display: none;'><a href="SingleCellExperiment.html#sliceAlternativeExperiments">sliceAlternativeExperiments</a></li><li data-type='method' style='display: none;'><a href="SingleCellExperiment.html#sliceAssays">sliceAssays</a></li><li data-type='method' style='display: none;'><a href="SingleCellExperiment.html#sliceReducedDimensions">sliceReducedDimensions</a></li></ul></li><li><a href="SummarizedExperiment.html">SummarizedExperiment</a><ul class='methods'><li data-type='method' style='display: none;'><a href="SummarizedExperiment.html#assay">assay</a></li><li data-type='method' style='display: none;'><a href="SummarizedExperiment.html#assayNames">assayNames</a></li><li data-type='method' style='display: none;'><a href="SummarizedExperiment.html#assays">assays</a></li><li data-type='method' style='display: none;'><a href="SummarizedExperiment.html#columnData">columnData</a></li><li data-type='method' style='display: none;'><a href="SummarizedExperiment.html#columnNames">columnNames</a></li><li data-type='method' style='display: none;'><a href="SummarizedExperiment.html#metadata">metadata</a></li><li data-type='method' style='display: none;'><a href="SummarizedExperiment.html#numberOfAssays">numberOfAssays</a></li><li data-type='method' style='display: none;'><a href="SummarizedExperiment.html#numberOfColumns">numberOfColumns</a></li><li data-type='method' style='display: none;'><a href="SummarizedExperiment.html#numberOfRows">numberOfRows</a></li><li data-type='method' style='display: none;'><a href="SummarizedExperiment.html#removeAssay">removeAssay</a></li><li data-type='method' style='display: none;'><a href="SummarizedExperiment.html#rowData">rowData</a></li><li data-type='method' style='display: none;'><a href="SummarizedExperiment.html#rowNames">rowNames</a></li><li data-type='method' style='display: none;'><a href="SummarizedExperiment.html#setAssay">setAssay</a></li><li data-type='method' style='display: none;'><a href="SummarizedExperiment.html#setAssayNames">setAssayNames</a></li><li data-type='method' style='display: none;'><a href="SummarizedExperiment.html#setAssays">setAssays</a></li><li data-type='method' style='display: none;'><a href="SummarizedExperiment.html#setColumnData">setColumnData</a></li><li data-type='method' style='display: none;'><a href="SummarizedExperiment.html#setColumnNames">setColumnNames</a></li><li data-type='method' style='display: none;'><a href="SummarizedExperiment.html#setMetadata">setMetadata</a></li><li data-type='method' style='display: none;'><a href="SummarizedExperiment.html#setRowData">setRowData</a></li><li data-type='method' style='display: none;'><a href="SummarizedExperiment.html#setRowNames">setRowNames</a></li><li data-type='method' style='display: none;'><a href="SummarizedExperiment.html#sliceAssays">sliceAssays</a></li></ul></li><li><a href="Vector.html">Vector</a><ul class='methods'><li data-type='method' style='display: none;'><a href="Vector.html#elementMetadata">elementMetadata</a></li><li data-type='method' style='display: none;'><a href="Vector.html#metadata">metadata</a></li><li data-type='method' style='display: none;'><a href="Vector.html#names">names</a></li><li data-type='method' style='display: none;'><a href="Vector.html#setElementMetadata">setElementMetadata</a></li><li data-type='method' style='display: none;'><a href="Vector.html#setMetadata">setMetadata</a></li><li data-type='method' style='display: none;'><a href="Vector.html#setNames">setNames</a></li></ul></li></ul><h3>Global</h3><ul><li><a href="global.html#CLONE">CLONE</a></li><li><a href="global.html#COMBINE">COMBINE</a></li><li><a href="global.html#COMBINE_COLUMNS">COMBINE_COLUMNS</a></li><li><a href="global.html#COMBINE_ROWS">COMBINE_ROWS</a></li><li><a href="global.html#flexibleCombineRows">flexibleCombineRows</a></li><li><a href="global.html#LENGTH">LENGTH</a></li><li><a href="global.html#NUMBER_OF_COLUMNS">NUMBER_OF_COLUMNS</a></li><li><a href="global.html#NUMBER_OF_ROWS">NUMBER_OF_ROWS</a></li><li><a href="global.html#presplitFactor">presplitFactor</a></li><li><a href="global.html#SLICE">SLICE</a></li><li><a href="global.html#SLICE_2D">SLICE_2D</a></li><li><a href="global.html#SPLIT">SPLIT</a></li><li><a href="global.html#which">which</a></li></ul>
</nav>

<div id="main">
    

    



    


    <section class="package">
        <h3> </h3>		
    </section>









    



    <section class="readme usertext">
        <article><h1>Bioconductor objects in Javascript</h1>
<p>This package aims to provide Javascript implementations of <a href="https://github.com/Bioconductor">Bioconductor</a> data structures for use in web applications.
Much like the original R code, we focus on the use of common generics to provide composability, allowing users to construct complex objects that &quot;just work&quot;.
We also attempt to circumvent Javascript's pass-by-reference behavior to avoid unintended modifications to unrelated objects when calling setter methods from their nested child objects.</p>
<h2>Quick start</h2>
<p>Here, we perform some generic operations on a <code>DataFrame</code> object, equivalent to Bioconductor's <code>S4Vectors::DFrame</code> class.</p>
<pre class="prettyprint source lang-js"><code>// Import using ES6 notation
import * as bioc from &quot;bioconductor&quot;;

// Construct a DataFrame
let results = new bioc.DataFrame(
    { 
        logFC: new Float64Array([-1, -2, 1.3, 2.1]),
        pvalue: new Float64Array([0.01, 0.02, 0.001, 1e-8])
    },
    {
        rowNames: [ &quot;p53&quot;, &quot;SNAP25&quot;, &quot;MALAT1&quot;, &quot;INS&quot; ]
    }
);

// Run generics
bioc.LENGTH(results);
bioc.SLICE(results, [ 2, 3, 1 ]); 
bioc.CLONE(results);

let more_results = new bioc.DataFrame(
    { 
        logFC: new Float64Array([0, 0.1, -0.1]),
        pvalue: new Float64Array([1e-5, 1e-4, 0.5])
    },
    {
        rowNames: [ &quot;GFP&quot;, &quot;mCherry&quot;, &quot;tdTomato&quot; ]
    }
);

bioc.COMBINE([results, more_results]);
</code></pre>
<p>See the <a href="https://ltla.github.io/bioconductor.js">reference documentation</a> for more details.</p>
<h1>Using generics</h1>
<p>Our generics allow users to operate on different objects in a consistent manner.
For example, a <code>DataFrame</code> allows us to store any object as a column as long as it defines methods for the <code>LENGTH</code>, <code>SLICE</code>, <code>CLONE</code> and <code>COMBINE</code> generics.
This enables the construction of complex objects like a <code>DataFrame</code> nested inside another <code>DataFrame</code>.</p>
<pre class="prettyprint source lang-js"><code>let genomic_results = new bioc.DataFrame(
    { 
        logFC: new Float64Array([-1, -2, 1.3, 2.1]),
        pvalue: new Float64Array([0.01, 0.02, 0.001, 1e-8]),
        location: new bioc.DataFrame({
            &quot;chromosome&quot;: [ &quot;chrA&quot;, &quot;chrB&quot;, &quot;chrC&quot;, &quot;chrD&quot; ],
            &quot;start&quot;: [ 1, 2, 3, 4 ],
            &quot;width&quot;: [ 10, 20, 30, 40 ],
            &quot;strand&quot;: new Uint8Array([-1, 1, 1, -1 ])
        })
    },
    {
        rowNames: [ &quot;p53&quot;, &quot;SNAP25&quot;, &quot;MALAT1&quot;, &quot;INS&quot; ]
    }
);

let subset = bioc.SLICE(genomic_results, { start: 2, end: 4 });
bioc.LENGTH(subset); 
subset.column(&quot;location&quot;);
</code></pre>
<p>Alternatively, we could store an <code>IRanges</code> (see below) as a column of our <code>DataFrame</code>.
All generics on the parent <code>DataFrame</code> will be automatically applied to the <code>IRanges</code> column.</p>
<pre class="prettyprint source lang-js"><code>let old_location = genomic_results.column(&quot;location&quot;);
let new_location = new bioc.GRanges(old_location.column(&quot;chromosome&quot;),
    new bioc.IRanges(old_location.column(&quot;start&quot;), old_location.column(&quot;width&quot;)),
    { strand: old_location.column(&quot;strand&quot;) });
genomic_results.setColumn(&quot;location&quot;, new_location, { inPlace: true });

subset = bioc.SLICE(genomic_results, { start: 2, end: 4 });
subset.column(&quot;location&quot;);
</code></pre>
<p>We mimic R's S4 generics using methods in Javascript classes.
For example, each vector-like class should define a <code>_bioconductor_LENGTH</code> method to quantify its concept of &quot;length&quot;.
The <code>LENGTH</code> function will then call this method to obtain a length value for any instance of any supported class.
We prefix this method with <code>_bioconductor_</code> to avoid collisions with other properties;
this allows safe monkey patching of third-party classes if they are sufficiently vector-like.</p>
<p>(Admittedly, the <code>LENGTH</code> function is not really necessary, as users could just call <code>_bioconductor_LENGTH</code> directly.
However, the latter is long and unpleasant to type, so we might as well wrap it in something that's easier to remember.
It would also require monkey patching of built-in classes like Arrays and TypedArrays, which is somewhat concerning as it risks interfering with the behavior of other packages.
By defining our own <code>LENGTH</code> function, we can safely handle the built-in classes as special cases without modifying their prototypes.)</p>
<h1>Mimicking copy-on-write</h1>
<p>We mimic R's copy-on-write behavior by returning a new object from any setter, rather than mutating the existing object.
This avoids silent pass-by-reference changes in separate objects, which would be particularly problematic in complex classes that contain many child objects.
In the example below, <code>another_reference</code> still retains the original set of row names while only <code>modified</code> has its row names removed.</p>
<pre class="prettyprint source lang-js"><code>// Construct a DataFrame
let results = new bioc.DataFrame(
    { 
        logFC: new Float64Array([-1, -2, 1.3, 2.1]),
        pvalue: new Float64Array([0.01, 0.02, 0.001, 1e-8])
    },
    {
        rowNames: [ &quot;p53&quot;, &quot;SNAP25&quot;, &quot;MALAT1&quot;, &quot;INS&quot; ]
    }
);

let another_reference = results;
let modified = results.setRowNames(null);
</code></pre>
<p>For users who are very sure that they are only operating on a single instance of the object,
or for those who wish to exploit pass-by-reference behavior to multiple multiple objects at once,
we can improve efficiency by requesting <code>inPlace: true</code> in the setter methods.</p>
<pre class="prettyprint source lang-js"><code>results.setRowNames(null, { inPlace: true });
another_reference.rowNames(); // this will now be null.
</code></pre>
<p>Note that this copy-on-write paradigm only applies to the setters defined in the <strong>bioconductor.js</strong> classes.
Assignments to base objects (e.g., arrays, TypedArrays) will still exhibit pass-by-reference behavior.
If there is a risk of inadvertently modifying a shared object, users should consider <code>CLONE</code>ing their object before modifying it.</p>
<pre class="prettyprint source lang-js"><code>// Returns a base object, i.e., Float64Array of log-fold changes.
let lfc = results.column(&quot;logFC&quot;);

// We clone it so that changes don't propagate to 'results' by reference.
// We can then apply our arbitrary modifications to the copy.
let lfc_copy = bioc.CLONE(lfc);
lfc_copy[0] = 100;

// Only 'more_modified' will contain the new log-FC's;
// 'results' itself is not affected.
let more_modified = results.setColumn(&quot;logFC&quot;, lfc_copy);
</code></pre>
<h1>Representing (genomic) ranges</h1>
<p>We can construct equivalents of Bioconductor's <code>IRanges</code> and <code>GRanges</code> objects, representing integer and genomic ranges respectively.
Similarly, Bioconductor's <code>GRangesList</code> is implemented as a <code>GroupedGRanges</code> in this package.</p>
<pre class="prettyprint source lang-js"><code>let ir = new bioc.IRanges(/* start = */ [1,2,3], /* width = */ [ 10, 20, 30 ]);
let gr = new bioc.GRanges([ &quot;chrA&quot;, &quot;chrB&quot;, &quot;chrC&quot; ], ir, { strand: [ 1, 0, -1 ] });

// Generics still work on these range objects:
bioc.LENGTH(gr);
bioc.SLICE(gr, [ 2, 1, 0 ]);
bioc.CLONE(gr);
</code></pre>
<p>We can find overlaps between two sets of ranges, akin to Bioconductor's <code>findOverlaps()</code> function:</p>
<pre class="prettyprint source lang-js"><code>let index = gr.buildOverlapIndex();
let gr2 = new bioc.GRanges([ &quot;chrA&quot;, &quot;chrC&quot;, &quot;chrA&quot; ], new bioc.IRanges([5, 3, 2], [9, 9, 9]));
let overlaps = index.overlap(gr2);
</code></pre>
<p>We can store per-range metadata in the <code>elementMetadata</code> field of each object, just like Bioconductor's <code>mcols()</code>.</p>
<pre class="prettyprint source lang-js"><code>let meta = gr.elementMetadata();
meta.setColumn(&quot;symbol&quot;, [ &quot;Nanog&quot;, &quot;Snap25&quot;, &quot;Malat1&quot; ], { inPlace: true });
gr.setElementMetadata(meta, { inPlace: true });
gr.elementMetadata().columnNames();
</code></pre>
<h1>Handling experimental assays</h1>
<p>The <code>SummarizedExperiment</code> object is a data structure for storing experimental data in a matrix-like object,
along with further annotations on the rows (usually features) and samples (usually columns).
To illustrate, let's mock up a small count matrix, ostensibly from an RNA-seq experiment:</p>
<pre class="prettyprint source lang-js"><code>// Making a column-major dense matrix of random data.
let ngenes = 100;
let nsamples = 20;
let expression = new Int32Array(ngenes * nsamples);
expression.forEach((x, i) => expression[i] = Math.random() * 10);
let mat = new bioc.DenseMatrix(ngenes, nsamples, expression);

// Mocking up row names, column annotations.
let rownames = [];
for (var g = 0; g &lt; ngenes; g++) {
    rownames.push(&quot;Gene_&quot; + String(g));
}

let treatment = new Array(nsamples);
treatment.fill(&quot;control&quot;, 0, 10);
treatment.fill(&quot;treated&quot;, 10, nsamples);
let sample_meta = new bioc.DataFrame({ group: treatment });
</code></pre>
<p>We can now store all of this information in a <code>SummarizedExperiment</code>:</p>
<pre class="prettyprint source lang-js"><code>let se = new bioc.SummarizedExperiment({ counts: mat }, 
    { rowNames: rownames, columnData: sample_meta });
</code></pre>
<p>This can be manipulated by generics for two-dimensional objects:</p>
<pre class="prettyprint source lang-js"><code>bioc.NUMBER_OF_ROWS(se);
bioc.SLICE_2D(se, { start: 0, end: 50 }, [0, 2, 4, 8, 10, 12, 14, 16, 18]);
bioc.COMBINE_COLUMNS([se, se]);
</code></pre>
<p>Similar implementations are provided for the <code>RangedSummarizedExperiment</code> and <a href="https://bioconductor.org/packages/SingleCellExperiment"><code>SingleCellExperiment</code></a> classes.</p>
<h1>Supported classes and generics</h1>
<p>For classes:</p>
<table>
<thead>
<tr>
<th><strong>Javascript</strong></th>
<th><strong>R/Bioconductor equivalent</strong></th>
</tr>
</thead>
<tbody>
<tr>
<td><a href="https://ltla.github.io/bioconductor.js/List.html"><code>List</code></a></td>
<td><code>base::list</code></td>
</tr>
<tr>
<td><a href="https://ltla.github.io/bioconductor.js/DataFrame.html"><code>DataFrame</code></a></td>
<td><code>S4Vectors::DFrame</code></td>
</tr>
<tr>
<td><a href="https://ltla.github.io/bioconductor.js/IRanges.html"><code>IRanges</code></a></td>
<td><code>IRanges::IRanges</code></td>
</tr>
<tr>
<td><a href="https://ltla.github.io/bioconductor.js/GRanges.html"><code>GRanges</code></a></td>
<td><code>GenomicRanges::GRanges</code></td>
</tr>
<tr>
<td><a href="https://ltla.github.io/bioconductor.js/GroupedGRanges.html"><code>GroupedGRanges</code></a></td>
<td><code>GenomicRanges::GRangesList</code></td>
</tr>
<tr>
<td><a href="https://ltla.github.io/bioconductor.js/SummarizedExperiment.html"><code>SummarizedExperiment</code></a></td>
<td><code>SummarizedExperiment::SummarizedExperiment</code></td>
</tr>
<tr>
<td><a href="https://ltla.github.io/bioconductor.js/RangedSummarizedExperiment.html"><code>RangedSummarizedExperiment</code></a></td>
<td><code>SummarizedExperiment::RangedSummarizedExperiment</code></td>
</tr>
<tr>
<td><a href="https://ltla.github.io/bioconductor.js/SingleCellExperiment.html"><code>SingleCellExperiment</code></a></td>
<td><code>SingleCellExperiment::SingleCellExperiment</code></td>
</tr>
</tbody>
</table>
<p>For generics:</p>
<table>
<thead>
<tr>
<th><strong>Javascript</strong></th>
<th><strong>R/Bioconductor equivalent</strong></th>
</tr>
</thead>
<tbody>
<tr>
<td><a href="https://ltla.github.io/bioconductor.js/LENGTH.html"><code>LENGTH</code></a></td>
<td><code>base::NROW</code></td>
</tr>
<tr>
<td><a href="https://ltla.github.io/bioconductor.js/SLICE.html"><code>SLICE</code></a></td>
<td><code>S4Vectors::extractROWS</code></td>
</tr>
<tr>
<td><a href="https://ltla.github.io/bioconductor.js/COMBINE.html"><code>COMBINE</code></a></td>
<td><code>S4Vectors::bindROWS</code></td>
</tr>
<tr>
<td><a href="https://ltla.github.io/bioconductor.js/CLONE.html"><code>CLONE</code></a></td>
<td>-</td>
</tr>
<tr>
<td><a href="https://ltla.github.io/bioconductor.js/NUMBER_OF_ROWS.html"><code>NUMBER_OF_ROWS</code></a></td>
<td><code>base::NROW</code></td>
</tr>
<tr>
<td><a href="https://ltla.github.io/bioconductor.js/NUMBER_OF_COLUMNS.html"><code>NUMBER_OF_COLUMNS</code></a></td>
<td><code>base::NCOL</code></td>
</tr>
<tr>
<td><a href="https://ltla.github.io/bioconductor.js/SLICE_2D.html"><code>SLICE_2D</code></a></td>
<td><code>base::&quot;[&quot;</code></td>
</tr>
<tr>
<td><a href="https://ltla.github.io/bioconductor.js/COMBINE_ROWS.html"><code>COMBINE_ROWS</code></a></td>
<td><code>S4Vectors::bindROWS</code></td>
</tr>
<tr>
<td><a href="https://ltla.github.io/bioconductor.js/COMBINE_COLUMNS.html"><code>COMBINE_COLUMNS</code></a></td>
<td><code>S4Vectors::bindCOLS</code></td>
</tr>
</tbody>
</table>
<h1>Further reading</h1>
<p>A high-level description of Bioconductor data structures is given in the <a href="https://doi.org/10.1038/nmeth.3252">&quot;Orchestrating high-throughput genomic analysis with Bioconductor&quot;</a> paper.</p>
<p>The formulation of the generics was mostly based on the code in the <a href="https://github.com/Bioconductor/S4Vectors"><strong>S4Vectors</strong></a> package.</p>
<p>The implementation of each class is based on the code in the corresponding R package, e.g., <code>GRanges</code> in <a href="https://bioconductor.org/packages/GenomicRanges"><strong>GenomicRanges</strong></a>.</p></article>
    </section>






    
    
</div>

<br class="clear">

<footer>
    Documentation generated by <a href="https://github.com/jsdoc3/jsdoc">JSDoc 3.6.11</a> on Sun Jul 06 2025 22:53:49 GMT+0000 (Coordinated Universal Time) using the <a href="https://github.com/clenemt/docdash">docdash</a> theme.
</footer>

<script>prettyPrint();</script>
<script src="scripts/polyfill.js"></script>
<script src="scripts/linenumber.js"></script>

<script src="scripts/search.js" defer></script>


<script src="scripts/collapse.js" defer></script>


</body>
</html>